src/calendarworker.cpp

```diff
@@ -319,11 +340,11 @@ void CalendarWorker::saveEvent(const CalendarData::Event &eventData, bool update
     save();
 }
 
-CalendarData::Event CalendarWorker::dissociateSingleOccurrence(const QString &uid, const QDateTime &recurrenceId)
+CalendarData::Event CalendarWorker::dissociateSingleOccurrence(const QString &instanceId, const QDateTime &recurrenceId)
 {
-    KCalendarCore::Event::Ptr event = mCalendar->event(uid);
-    if (!event) {
-        qWarning("Event to create occurrence replacement for not found");
+    KCalendarCore::Incidence::Ptr event = mCalendar->instance(instanceId);
+    if (!event || event->hasRecurrenceId()) {
```

Comment:
Is this proper? So coming eventually from calendareventmodification. Should be possible to have an exception on recurring exception (i.e. repeats and has recurrence id due to replacing something earlier)? The old code allowed such so should this get the parent event if recurring?

Reply:
There is no support at the moment for recurring exceptions in KCalendarCore, as far as I know.

At the moment in the UI, there is no way to make an exception out of an exception. In the sense that a CalendarEventModification that call ::dissociateSingleOccurrence() can be called only when modifying a recurring parent and selecting "modify this occurrence only".

Previously, the code allowed was adding an exception to a recurring parent (only the uid was fetch, without recurrenceId). So here the code is still doing the same, checking that the instanceId was not belonging to an exception.

That being said, to get the exact same behaviour (the caller is an exception, but only pass the uid), I can change for the following:
```
KCalendarCore::Incidence::Ptr event = mCalendar->instance(instanceId);
if (event && event->hasRecurrenceId()) {
    event = mCalendar->incidence(event->uid());
}
if (!event) {
    qWarning(...
```

I thought it was a bit verbose for a case that is not possible from UI design neither from internal KCalendarCore implementation.

Reply:
At least we've disabled recurrence on the jolla-calendar side for exceptions, but I'm not sure anymore was that to keep it just simpler or were there deeper problems. But if we do e.g. that "modify this and later" to change time of a meeting it should be done by modifying the exception and needs repetition?

Reply:
The this and later modifications, still didn't make in. But it's a good argument.

The implementation of the this and later does not require to dissociate from an exception. At least not the way it is implemented in KCalendarCore (it may be broken though, don't know). The exception is always dissociated from the parent. Actually, dissociation means internally "provide a recurrenceId that match one of the occurrence of the _parent_" and not one of the occurrence of the current this and later exception.

So let say you have a recurring parent P, recurring every day at noon.
- Then at one point in time, you decide that it should start recurring every day at 1pm. It means to create a non recurring clone of P and `setRecurrenceId(noon on day D1)`, `setThisAndFuture(true)` and `setDtStart(1pm on day D1)`.
- Later, you decide to move it every day at 11am. It means to create a non recurring clone of P and `setRecurrenceId(noon on day D2)`, `setThisAndFuture(true)` and `setDtStart(11am on day D2)`.

So the call to `dissociateSingleOccurrence()` must always be called with the `instanceId` of the parent.